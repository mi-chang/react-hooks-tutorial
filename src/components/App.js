import React, { useReducer, useEffect } from 'react';
import '../App.css';
import Header from './Header';
import Movie from './Movie';
import Search from './Search';

const API_KEY = process.env.REACT_APP_API_KEY;

const initialState = {
  loading: true,
  movies: [],
  errorMessage: null
};

const reducer = (state, action) => {
  switch (action.type) {
    case 'SEARCH_MOVIE_REQUEST':
      return {
        ...state,
        loading: true,
        errorMessage: null
      };
    case 'SEARCH_MOVIE_SUCCESS':
      return {
        ...state,
        loading: false,
        movies: action.payload
      };
    case 'SEARCH_MOVIE_FAILURE':
      return {
        ...state,
        loading: false,
        errorMessage: action.error
      };
    default:
      return state;
  }
};

const App = () => {
  const [state, dispatch] = useReducer(reducer, initialState);

  useEffect(() => {
    fetch(`https://www.omdbapi.com/?s=man&apikey=${API_KEY}`)
    .then(res => res.json())
    .then(json => {
      dispatch({
        type: 'SEARCH_MOVIE_SUCCESS',
        payload: json.Search
      });
    });
  }, []);

  const search = searchValue => {
    dispatch({
      type: 'SEARCH_MOVIE_REQUEST'
    });

    fetch(`https://www.omdbapi.com/?s=${searchValue}&apikey=${API_KEY}`)
    .then(res => res.json())
    .then(json => {
      if (json.Response === 'True') {
        dispatch({
          type: 'SEARCH_MOVIE_SUCCESS',
          payload: json.Search
        });
      } else {
        dispatch({
          type: 'SEARCH_MOVIE_FAILURE',
          error: json.Error
        })
      }
    });
  };

  const { movies, errorMessage, loading } = state;

  return (
    <div className="App">
      <Header text='HOOKED' />
      <Search search={search} />
      <p className='App-intro'>Sharing a few our favourite movies</p>
      <div className='movies'>
        {loading && !errorMessage ? (
          <span>loading...</span>
        ) : errorMessage ? (
          <div className='errorMessage'>{errorMessage}</div>
        ) : (
          movies.map((movie, index) => (
            <Movie key={`${index}-${movie.Title}`} movie={movie} />
            ))
        )}
      </div>
    </div>
  );
};

export default App;
