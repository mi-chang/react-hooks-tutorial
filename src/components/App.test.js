import React from 'react';
import { render, act, fireEvent } from '@testing-library/react';
import App from './App';

describe('App', () => {
  it('shows loading before request', () => {
    const { container, getByText } = render(<App />);
 
    expect(getByText('Sharing a few our favourite movies')).toBeInTheDocument();
    expect(getByText('loading...')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  })

  it('searches "man" at first', async () => {
    const fetchAPI = jest.spyOn(global, 'fetch').mockImplementation(() =>
      Promise.resolve({
        json: () => Promise.resolve({
          Response: 'True',
          Search: [
            {
              Title: 'man',
              Poster: 'http://some.com/image/man',
              Year: 1111
            }
          ]
        })
      })
    );
    
    let myContainer, myGetByText, myGetByRole, myGetByAltText;
    await act(async () => {
      const { container, getByText, getByRole, getByAltText }
        = render(<App />);
      // TODO Better way must exist.
      [myContainer, myGetByText, myGetByRole, myGetByAltText]
        = [container, getByText, getByRole, getByAltText];
    });

    expect(myGetByText('Sharing a few our favourite movies')).toBeInTheDocument();
    expect(myGetByText('man')).toBeInTheDocument();
    expect(myGetByText('(1111)')).toBeInTheDocument();
    expect(myGetByAltText('The movie titled: man')).toBeInTheDocument();
    expect(myGetByRole('img').src).toBe('http://some.com/image/man');
    expect(fetchAPI).toBeCalledTimes(1);
    expect(fetchAPI.mock.calls[0][0]).toContain('s=man');
    expect(myContainer).toMatchSnapshot();

    fetchAPI.mockRestore();
  });

  it('searches "woman"', async () => {
    const fetchAPI = jest.spyOn(global, 'fetch').mockImplementation(() =>
      Promise.resolve({
        json: () => Promise.resolve({
          Response: 'True',
          Search: [
            {
              Title: 'man',
              Poster: 'http://some.com/image/man',
              Year: 1111
            }
          ]
        })
      })
    );
    
    let myContainer, myGetByLabelText, myGetByText, myGetByRole, myGetByAltText;
    await act(async () => {
      const { container, getByLabelText, getByText, getByRole, getByAltText }
        = render(<App />);
      // TODO Better way must exist.
      [myContainer, myGetByLabelText, myGetByText, myGetByRole, myGetByAltText]
        = [container, getByLabelText, getByText, getByRole, getByAltText];
    });
    expect(fetchAPI).toBeCalledTimes(1);

    fetchAPI.mockImplementation(() =>
      Promise.resolve({
        json: () => Promise.resolve({
          Response: 'True',
          Search: [
            {
              Title: 'woman',
              Poster: 'http://some.com/image/woman',
              Year: 2222
            }
          ]
        })
      })
    );
    fireEvent.change(myGetByLabelText('Title'), { target: { value: 'woman' } });
    await act(async () => fireEvent.click(myGetByText('SEARCH')));

    expect(myGetByText('Sharing a few our favourite movies')).toBeInTheDocument();
    expect(myGetByText('woman')).toBeInTheDocument();
    expect(myGetByText('(2222)')).toBeInTheDocument();
    expect(myGetByAltText('The movie titled: woman')).toBeInTheDocument();
    expect(myGetByRole('img').src).toBe('http://some.com/image/woman');
    expect(fetchAPI).toBeCalledTimes(2);
    expect(fetchAPI.mock.calls[1][0]).toContain('s=woman');
    expect(myContainer).toMatchSnapshot();

    fetchAPI.mockRestore();
  });

  it('shows error message', async () => {
    const fetchAPI = jest.spyOn(global, 'fetch').mockImplementation(() =>
      Promise.resolve({
        json: () => Promise.resolve({
          Response: 'True',
          Search: [
            {
              Title: 'aaa',
              Poster: 'http://some.com/image/aaa',
              Year: 1111
            }
          ]
        })
      })
    );
    
    let myContainer, myGetByLabelText, myGetByText;
    await act(async () => {
      const { container, getByLabelText, getByText } = render(<App />);
      // TODO Better way must exist.
      [myContainer, myGetByLabelText, myGetByText] = [container, getByLabelText, getByText];
    });
    expect(fetchAPI).toBeCalledTimes(1);

    fetchAPI.mockImplementation(() =>
      Promise.resolve({
        json: () => Promise.resolve({
          Response: 'False',
          Error: 'Somethig bad!'
        })
      })
    );
    fireEvent.change(myGetByLabelText('Title'), { target: { value: 'abc' } });
    await act(async () => fireEvent.click(myGetByText('SEARCH')));

    expect(myGetByText('Sharing a few our favourite movies')).toBeInTheDocument();
    expect(myGetByText('Somethig bad!')).toBeInTheDocument();
    expect(fetchAPI).toBeCalledTimes(2);
    expect(fetchAPI.mock.calls[1][0]).toContain('s=abc');
    expect(myContainer).toMatchSnapshot();

    fetchAPI.mockRestore();
  });
});
