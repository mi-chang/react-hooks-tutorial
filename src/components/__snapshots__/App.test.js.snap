// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App searches "man" at first 1`] = `
<div>
  <div
    class="App"
  >
    <header
      class="App-header"
    >
      <h2>
        HOOKED
      </h2>
    </header>
    <form
      class="search"
    >
      <label
        for="searchText"
      >
        Title
      </label>
      <input
        id="searchText"
        type="text"
        value=""
      />
      <input
        type="submit"
        value="SEARCH"
      />
    </form>
    <p
      class="App-intro"
    >
      Sharing a few our favourite movies
    </p>
    <div
      class="movies"
    >
      <div
        class="movie"
      >
        <h2>
          man
        </h2>
        <div>
          <img
            alt="The movie titled: man"
            src="http://some.com/image/man"
            width="200"
          />
        </div>
        <p>
          (
          1111
          )
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`App searches "woman" 1`] = `
<div>
  <div
    class="App"
  >
    <header
      class="App-header"
    >
      <h2>
        HOOKED
      </h2>
    </header>
    <form
      class="search"
    >
      <label
        for="searchText"
      >
        Title
      </label>
      <input
        id="searchText"
        type="text"
        value=""
      />
      <input
        type="submit"
        value="SEARCH"
      />
    </form>
    <p
      class="App-intro"
    >
      Sharing a few our favourite movies
    </p>
    <div
      class="movies"
    >
      <div
        class="movie"
      >
        <h2>
          woman
        </h2>
        <div>
          <img
            alt="The movie titled: woman"
            src="http://some.com/image/woman"
            width="200"
          />
        </div>
        <p>
          (
          2222
          )
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`App shows error message 1`] = `
<div>
  <div
    class="App"
  >
    <header
      class="App-header"
    >
      <h2>
        HOOKED
      </h2>
    </header>
    <form
      class="search"
    >
      <label
        for="searchText"
      >
        Title
      </label>
      <input
        id="searchText"
        type="text"
        value=""
      />
      <input
        type="submit"
        value="SEARCH"
      />
    </form>
    <p
      class="App-intro"
    >
      Sharing a few our favourite movies
    </p>
    <div
      class="movies"
    >
      <div
        class="errorMessage"
      >
        Somethig bad!
      </div>
    </div>
  </div>
</div>
`;

exports[`App shows loading before request 1`] = `
<div>
  <div
    class="App"
  >
    <header
      class="App-header"
    >
      <h2>
        HOOKED
      </h2>
    </header>
    <form
      class="search"
    >
      <label
        for="searchText"
      >
        Title
      </label>
      <input
        id="searchText"
        type="text"
        value=""
      />
      <input
        type="submit"
        value="SEARCH"
      />
    </form>
    <p
      class="App-intro"
    >
      Sharing a few our favourite movies
    </p>
    <div
      class="movies"
    >
      <span>
        loading...
      </span>
    </div>
  </div>
</div>
`;
